### PR TITLE
#47689 improve run-speed of pip package state 

### PR DESCRIPTION
**Pull request by Orlando-Richards**
### What does this PR do?

Only load the current package list once when checking multiple packages, rather than once for each package.

### What issues does this PR fix or reference?

#47689 

### Previous Behavior

pip freeze is executed once for each package in a list when checking multiple pip packages.

### New Behavior

pip freeze is only executed once, and the result is re-used for each package check.

### Tests written?

No

### Commits signed with GPG?

No

### CODE REVIEW

**Comment by gtmanfred** [comment]
There are some lint failures that will need to be cleaned up https://jenkins.saltstack.com/job/PR/job/salt-pr-docs-n/17344/violations/

**Comment by OrlandoArcapix** [comment]
Lint errors sorted

**Review by gtmanfred — APPROVED** [review]

File: salt/states/pip_state.py [hunk 1]
```diff
@@ -186,16 +186,19 @@ def _check_pkg_version_format(pkg):
 
 def _check_if_installed(prefix, state_pkg_name, version_spec, ignore_installed,
                         force_reinstall, upgrade, user, cwd, bin_env, env_vars,
-                        **kwargs):
+                        pip_list=False, **kwargs):
```
Comment (cachedout) [inline salt/states/pip_state.py]: IMHO we need to have some documentation about what this kwarg does and the type of data expected here.

**Review by cachedout — CHANGES_REQUESTED** [review]
Some docs, please

File: salt/states/pip_state.py (hunk 1 above)
Comment (OrlandoArcapix) [reply salt/states/pip_state.py]: I was surprised by that kwargs too! It wasn't in the release code I initially patched on my systems, but it was there when I pulled the 2017.7 branch to do the pull request against.

Looks like it came from here: https://github.com/saltstack/salt/commit/7c46d9d0d457a686afb12860a655c0ecc15e427b

Comment (rallytime) [reply salt/states/pip_state.py]: @OrlandoArcapix I think we just want some documentation for your new kwarg, `pip_list`. I know the terminology was a little confusing there. :)

**Comment by OrlandoArcapix** [comment]
Ahh, I see @rallytime ! 

Not sure where to put that - the extra argument I've added is only used in an internal function, and it's arguments are not currently documented anywhere that I can see - can you point me to the right place to add the extra argument info?

```
def _check_if_installed(prefix, state_pkg_name, version_spec, ignore_installed,
                        force_reinstall, upgrade, user, cwd, bin_env, env_vars,
                        pip_list=False, **kwargs):
    # result: None means the command failed to run
    # result: True means the package is installed
    # result: False means the package is not installed
```

**Comment by rallytime** [comment]
@OrlandoArcapix Ah, you're right. I misread the code and thought it was an exposed variable, but it is not. I suppose it would be nice to have it documented somewhere maybe inline what the point of the variable is? I know none of the other kwargs are documented in `_check_if_installed`, but adding this would be a good place to start.

You can either do a doc string, or just a comment. The doc string won't populate the API documentation, but at least there will be something in the code to reference.

**Comment by OrlandoArcapix** [comment]
Docblock added with new kwarg described!

**Comment by rallytime** [comment]
Thanks @OrlandoArcapix! There are some lint errors on this now: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/22663/violations/file/salt/states/pip_state.py/

Can you fix those up?

**Comment by OrlandoArcapix** [comment]
Oops - sorry! All sorted.

**Comment by rallytime** [comment]
@OrlandoArcapix This change is failing the following test:

- integration.states.test_pip_state.PipStateTest.test_46127_pip_env_vars

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23706/

Can you take a look?

**Comment by rallytime** [comment]
Hi @OrlandoArcapix - Just a bump here. Were you able to take a look at that failing test?

**Comment by OrlandoArcapix** [comment]
Hi @rallytime, sorry, I've not had much of a chance to dig into it. I had a brief look at first, and I thought it might just need a change to the string checks in the tests. I'd worry that there was something more fundamental though, in case the test really is flagging a bad behaviour - I'm not familiar with the pip module code to know for sure that my quick fix was sensible!

Having said that, I've been using it in production now for hundreds of system builds and it hasn't let me down yet!

I can have a shot at changing the test's string check if you think that's a good way to proceed  - let me know.

**Comment by gtmanfred** [comment]
As long as the test still is valid, that should be fine to change the test string.

**Comment by OrlandoArcapix** [comment]
I've updated the test to match the string currently returned. 

For what it's worth, I have been using it for a couple of months now, across a lot of different pip installation situations, and it's been working fine - but I wouldn't say no to someone with knowledge of the code workings doing a review of the change!

**Review by rallytime — APPROVED** [review]

**Pull request merged**